toolchain/check/thunk.cpp

```diff
@@ -0,0 +1,421 @@
+                         .kind = is_generic ? SemIR::BindSymbolicName::Kind
+                                            : SemIR::BindName::Kind,
+                         .type_id = inner_type_id,
+                         .entity_name_id = entity_name_id,
+                         .value_id = SemIR::InstId::None}));
+    if (is_generic) {
+      context.scope_stack().PushCompileTimeBinding(bind_name_id);
+    }
+
+    auto type_expr_region_id = context.sem_ir().expr_regions().Add(
+        {.block_ids = {SemIR::InstBlockId::Empty},
+         .result_id = context.types().GetInstId(inner_type_id)});
+    bool inserted = context.bind_name_map()
+                        .Insert(new_pattern_id,
+                                {.bind_name_id = bind_name_id,
+                                 .type_expr_region_id = type_expr_region_id})
+                        .is_inserted();
+    CARBON_CHECK(inserted);
+  } else if (auto return_slot = pattern.TryAs<SemIR::ReturnSlotPattern>()) {
+    new_pattern_id = AddPatternInst(
+        context,
+        SemIR::LocIdAndInst::UncheckedLoc(
+            SemIR::LocId(pattern_id),
+            SemIR::ReturnSlotPattern{.type_id = inner_type_id,
+                                     .type_inst_id = SemIR::TypeInstId::None}));
+  } else {
+    CARBON_CHECK(pattern.Is<SemIR::ErrorInst>(),
+                 "Unexpected pattern {0} in function signature", pattern);
+    return SemIR::ErrorInst::InstId;
+  }
+
+  // Rebuild parameter.
+  if (param) {
+    auto type_id =
+        SemIR::GetTypeOfInstInSpecific(context.sem_ir(), specific_id, param_id);
+    new_pattern_id = AddPatternInst(
+        context,
+        SemIR::LocIdAndInst::UncheckedLoc(
+            SemIR::LocId(param_id),
+            SemIR::AnyParamPattern{.kind = param->kind,
+                                   .type_id = type_id,
+                                   .subpattern_id = new_pattern_id,
+                                   .index = SemIR::CallParamIndex::None}));
+  }
+
+  // Rebuild `addr`.
+  if (addr) {
+    auto type_id =
+        SemIR::GetTypeOfInstInSpecific(context.sem_ir(), specific_id, addr_id);
+    new_pattern_id = AddPatternInst(
+        context, SemIR::LocIdAndInst::UncheckedLoc(
+                     SemIR::LocId(addr_id),
+                     SemIR::AddrPattern{.type_id = type_id,
+                                        .inner_id = new_pattern_id}));
+  }
+
+  return new_pattern_id;
+}
+
+static auto ClonePatternBlock(Context& context, SemIR::SpecificId specific_id,
+                              SemIR::InstBlockId inst_block_id)
+    -> SemIR::InstBlockId {
+  if (!inst_block_id.has_value()) {
+    return SemIR::InstBlockId::None;
+  }
+  auto orig_block = context.inst_blocks().Get(inst_block_id);
+
+  llvm::SmallVector<SemIR::InstId> block;
+  block.reserve(orig_block.size());
+  for (auto inst_id : orig_block) {
+    block.push_back(ClonePattern(context, specific_id, inst_id));
+  }
```

Comment:
I feel like this is pretty similar to code I was recently getting comments on; maybe something like:
```suggestion
  llvm::SmallVector<SemIR::InstId> block(llvm::map_range(orig_block, [&](SemIR::InstId inst_id) { return ClonePattern(context, specific_id, inst_id)}));
```